### ISSUE TITLE
BBS spk security against an untrusted/compromised issuer

### ISSUE THREAD

**Issue by BasileiosKal** [decentralized-identity/bbs-signature#75]
Let’s say that the issuer gets compromised, i.e., both their secret key `x` and the records with the signatures they signed will become known. Let’s assume a holder has a bbs+ signature `(A, e, s)` from that issuer with `A = B * (1/(e+x))` where `B = P1 + H0*s + H_1*msg_1 + … + H_L*msg_L`. Any spk will be of the form, `spk = (A', Abar, D, ...)` where 
```
A' = A * r1 =  B * r1 * (1/(e+x))
```
 and 
```
Abar = A' * x =  B * r1 * (x/(e+x))
```
The values `(B, B*x, A', Abar)` essentially form a DDH tuple, which is easy to solve using pairings. Specifically, it will hold, 
```
e(B*x, A’) = e(B*x, B*r1*(1/(e+x))) = e(B, B) ^(r1*(x/(e+x))) = e(B, B*r1*(x/(e+x)))
```
meaning that,
```
e(B*x, A') = e(B, Abar)
```
So, if an adversary who knows `x` and `s`, sees (or has already seen) a proof with `A’` and `Abar`, they can just try different messages, construct `B` and check the above equality between the pairings. Since the messages will not be many this is not hard to do and will result in the secret messages of the holder been revealed. 

If the Issuer is untrusted, they can use the above equation to track the holder by correlating the spk's with specific signatures.

This is mitigated a little by using blind or bound signatures. If the commitment is not revealed an adversary will not be able to construct `B` by trying possible hidden messages. However, it is reasonable to assume here that if the issuers secret key and the signatures they signed are revealed so will the commitments (they will be known -and possibly recorded- by the issuer). Furthermore, if the issuer is untrusted. they can still correlate the holder.

As a result, the bbs+ proofs are not secure against an untrusted or compromised issuer. This is not that much of an issue and is like what happens in most PKIs but it might be worth documenting it in the security considerations in the spec so we will have an accurate security model.

**Comment by tplooker** [comment]
Really interesting issue, IMO this definitely needs to be in the security considerations section.

**Comment by tplooker** [comment]
> So, if an adversary who knows x and s, sees (or has already seen) a proof with A’ and Abar, they can just try different messages, construct B and check the above equality between the pairings. Since the messages will not be many this is not hard to do and will result in the secret messages of the holder been revealed.

One potential mitigation we should look to include here is for issuers to not keep a copy of signatures they issue? So if the issuer is compromised chances are the attacker doesnt get a copy of the s value for a signature?

**Comment by BasileiosKal** [comment]
Yes, that's a good point. An honest issuer should discard at the very least `s` and not keep a record of it. If an adversary doesn't have `s`, even if they have the issuer's secret key, the above doesn't work.

**Comment by tplooker** [comment]
Yes agreed, because `s` is apart of the original signature over all messages and the primary attack vector here is being able to reveal, un-revealed messages in a proof of a signature involving `s` chances are that if the issuers private key is compromised and so is `s` the attacker probably also got a copy of the original messages signed by the signature. None the less we should document this consideration carefully in the spec so implementers are aware.

**Comment by tplooker** [comment]
I've tagged this issue as ready for PR as it appears we are clear around the resolution which is to add a note to the security considerations section

**Comment by andrewwhitehead** [comment]
I'm having trouble seeing what correlation is enabled here.

    e(B*x, A') = e(B, Abar)

This seems self-evident given that `Abar = A'x`, plus you could substitute any point for `B` in the equation and the result would be the same.

**Comment by BasileiosKal** [comment]
That is absolutely correct. Tbo kinda annoyed with my self that i missed it. I'm really sorry about that. Although It does seem to me that the issuer may still find a relationship that would correlate the prover. For example, they could calculate `K = e(p1*c, B*r1)` (since they know e, x, d) and check `e(p1*c, A') = K*(1/x+e)`. Another thing i missed though is that any method like that works on type 2 or 3 pairings. However we are exclusively using type 3 since we are using the construction of CDL16. Maybe just a note documenting that will suffice instead of the addition in #86.

**Comment by tplooker** [comment]
Just re-opening this issue so we can be sure to capture any updated text required in the security considerations merged in #86

**Comment by BasileiosKal** [comment]
Addressed in #91. Closing
